### PR TITLE
Fix invoking cargo without consulting CARGO env var or standard installation paths

### PR DESCRIPTION
**Pull request by Veetaha**
Followup for #4329

The pr essentially fixes [this bug](https://youtu.be/EzQ7YIIo1rY?t=2189)

cc @lefticus

### CODE REVIEW

File: crates/rust-analyzer/src/main_loop/handlers.rs [hunk 1]
```diff
@@ -1031,10 +1032,15 @@ fn to_lsp_runnable(
             target.map_or_else(|| "run binary".to_string(), |t| format!("run {}", t))
         }
     };
+    let cargo_path = ra_toolchain::cargo()
+        .to_str()
+        .context("Path to cargo executable contains invalid UTF8 characters")?
```
Comment (lnicola) [inline crates/rust-analyzer/src/main_loop/handlers.rs:1037]:
```suggestion
        .context("Path to `cargo` executable contains invalid UTF8 characters")?
```

Comment (Veetaha) [reply crates/rust-analyzer/src/main_loop/handlers.rs]: Done, thanks!

**Comment by Veetaha** [comment]
@matklad this should be good to go

**Comment by matklad** [comment]
Hm, I am not sure if this is conceptually the right fix here, I think we should move towards keeping only semantic info on the server, and replacing the `bin` with something like `type`.

**Comment by Veetaha** [comment]
It's not that bad since this is `rust-analyzer` crate in question and not something pure like `ra_hir` or `ra_ide`, but do you really want to move more code to the frontend :D?

**Comment by Veetaha** [comment]
I changed Runnable.bin -> Runnable.kind (`type` would be ugly because it is a non-contextual keyword in Rust).
@matklad this is ready to merge after your approval

**Comment by Veetaha** [comment]
boop @matklad

**Comment by matklad** [comment]
bors r+

I'll tweak the proto stuff myself afterwards though ;-)

**Pull request merged**